LocalLang/Semantics.lean

```diff
@@ -1,12 +1,41 @@
 import LocalLang.AST
 
+inductive Ctx : Type where
+  | hole : Ctx
+  | bin_op_lhs : Ctx → BinOp → Expr → Ctx
+  | bin_op_rhs : ℕ → BinOp → Ctx → Ctx
+  | fun_call : String → List ℕ → Ctx → List Expr → Ctx
+
+def Ctx.fill (e : Expr) : Ctx → Expr
+  | hole => e
+  | bin_op_lhs ctx op e' => Expr.binOp op (ctx.fill e) e'
+  | bin_op_rhs n op ctx => Expr.binOp op (.const n) (ctx.fill e)
+  | fun_call f es_left ctx es_right =>
+      Expr.funCall f ((.const ·) <$> es_left ++ [ctx.fill e] ++ es_right)
+
+abbrev VarContext := Std.HashMap String ℕ
+
 -- TODO: define
-inductive SmallStep (defs : Definitions) : Expr → Expr → Prop where
+inductive SmallStep (defs : Definitions) : (VarContext × Expr) → (VarContext × Expr) → Prop where
+  | var_step : V[x]? = some n →
+      SmallStep defs (V, (.var x)) (V, .const n)
+  | bin_op_step {op : BinOp} :
+      SmallStep defs (V, .binOp op (.const e₁) (.const e₂)) (V, .const <| op.eval e₁ e₂)
+  | ctx_step (ctx : Ctx) : SmallStep defs (V, e₁) (V, e₂) →
+      SmallStep defs (V, ctx.fill e₁) (V, ctx.fill e₂)
+  | fun_step {es : List ℕ} {func : Function} (Hf : defs.find? (·.name = f) = some func)
+      (Hpn : func.parameters.length = es.length):
+      SmallStep defs
+        (V, .funCall f <| (.const ·) <$> es)
+        (.ofList <| V.toList ++ func.parameters.zip es, func.body)
```

Comment:
This does not work.  Consider the following:

```
f x = g (x + 1) + x
g x = x
f 0
```

According to your small step semantics this evaluates to 2, but it should evaluate to 1.